node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -428,31 +366,67 @@ async fn distribute_collation<Context>(
 	gum::debug!(
 		target: LOG_TARGET,
 		para_id = %id,
-		relay_parent = %relay_parent,
-		candidate_hash = ?receipt.hash(),
+		candidate_relay_parent = %candidate_relay_parent,
+		relay_parent_mode = ?relay_parent_mode,
+		candidate_hash = ?candidate_hash,
 		pov_hash = ?pov.hash(),
 		core = ?our_core,
 		?current_validators,
 		"Accepted collation, connecting to validators."
 	);
 
+	let validators_at_relay_parent = &mut per_relay_parent.validator_group.validators;
+	if validators_at_relay_parent.is_empty() {
+		*validators_at_relay_parent = current_validators.validators.clone();
+	}
+
 	// Issue a discovery request for the validators of the current group:
+	//
+	// TODO [now]: some kind of connection management is necessary to avoid
+	// dropping peers from e.g. implicit view assignments.
 	connect_to_validators(ctx, current_validators.validators.into_iter().collect()).await;
 
-	state.our_validators_groups.insert(relay_parent, ValidatorGroup::new());
-
 	if let Some(result_sender) = result_sender {
-		state.collation_result_senders.insert(receipt.hash(), result_sender);
+		state.collation_result_senders.insert(candidate_hash, result_sender);
 	}
 
-	state
-		.collations
-		.insert(relay_parent, Collation { receipt, pov, status: CollationStatus::Created });
+	per_relay_parent.collations.insert(
+		candidate_hash,
+		Collation { receipt, parent_head_data_hash, pov, status: CollationStatus::Created },
+	);
+
+	// If prospective parachains are disabled, a leaf should be known to peer.
+	// Otherwise, it should be present in allowed ancestry of some leaf.
+	//
+	// It's collation-producer responsibility to verify that there exists
+	// a hypothetical membership in a fragment tree for candidate.
+	let interested =
+		state
+			.peer_data
+			.iter()
+			.filter(|(_, PeerData { view: v, .. })| match relay_parent_mode {
+				ProspectiveParachainsMode::Disabled => v.contains(&candidate_relay_parent),
+				ProspectiveParachainsMode::Enabled => v.iter().any(|block_hash| {
+					state
+						.implicit_view
+						.known_allowed_relay_parents_under(block_hash, Some(id))
+						.unwrap_or_default()
+						.contains(&candidate_relay_parent)
+				}),
+			});
 
-	let interested = state.peers_interested_in_leaf(&relay_parent);
 	// Make sure already connected peers get collations:
-	for peer_id in interested {
-		advertise_collation(ctx, state, relay_parent, peer_id).await;
+	for (peer_id, peer_data) in interested {
+		advertise_collation(
```

Comment:
Reviewed that this is only advertising to peers with the relay-parent in their implicit view.